src/metabase/lib/query.cljc

```diff
@@ -433,3 +434,73 @@
   (-> a-query
       (dissoc a-query :lib/metadata)
       lib.cache/discard-query-cache))
+
+(defn- stage-seq* [dataset-query]
+  (cond
+    (vector? dataset-query)
+    (case (first dataset-query)
+      :metric
+      [{:source-card (get dataset-query 2)}]
+
+      (mapcat stage-seq* dataset-query))
+
+    (map? dataset-query)
+    (concat (:stages dataset-query) (mapcat stage-seq* (vals dataset-query)))
+
+    :else
+    []))
+
+(defn- stage-seq [card-id dataset-query]
+  (map #(assoc % ::from-card card-id) (stage-seq* dataset-query)))
+
+(defn- expand-stage [metadata-provider stage]
+  (let [card-id (:source-card stage)
+        card (when card-id (lib.metadata/card metadata-provider card-id))
+        expanded-query (some->> card
+                                :dataset-query
+                                (query metadata-provider))]
+    (stage-seq card-id expanded-query)))
+
+(defn- add-stage-dep [graph stage]
+  (let [card-id  (:source-card  stage)
+        table-id (:source-table stage)
+        from-id  (::from-card   stage)]
+    (cond-> graph
+      card-id  (dep/depend [:card from-id] [:card  card-id])
+      table-id (dep/depend [:card from-id] [:table table-id]))))
+
+(defn- build-graph [graph source-id metadata-provider dataset-query]
```

Comment:
Here and in both functions below:
- `dataset-query` means "nearly raw from appDB, needs `normalize`"
- `legacy-query` means "modern, nicely formed legacy MBQL"
- `query` means "pMBQL query in lib style"

So these functions should all be using `query`.

Reply:
Does it matter that it's shadowing the function called `query`?